src/dmd/clone.d

```diff
@@ -521,9 +521,9 @@ FuncDeclaration buildOpEquals(StructDeclaration sd, Scope* sc)
 
 /******************************************
  * Build __xopEquals for TypeInfo_Struct
- *      static bool __xopEquals(ref const S p, ref const S q)
+ *      bool __xopEquals(ref const S p)
  *      {
- *          return p == q;
+ *          return p == this;
```

Comment:
I've noticed that the order is reversed in druntime, calling `rhs.__xopEquals(lhs)`, and here it's reversed back again. But an aggregate defining an `extern(D) bool opEquals(const ref T rhs)` does not need the generated `__xopEquals`, so the non-wrapped version gets called with `rhs.opEquals(lhs)` order.

Have you tried to simplify this without any swapping and then hit issues? I'm asking because I suspect this weirdness to be the culprit for remaining issues in https://github.com/ldc-developers/ldc/pull/3873 - the whole testsuite is green, but a compiler built itself with the new ABI then miscompiles, complaining about template forward references IIRC. I suspect something in the DMD frontend to depend on this swapped `opEquals` order.

Reply:
> I've noticed that the order is reversed in druntime, calling rhs.__xopEquals(lhs), and here it's reversed back again. But an aggregate defining an extern(D) bool opEquals(const ref T rhs) does not need the generated __xopEquals, so the non-wrapped version gets called with rhs.opEquals(lhs) order.

Can't say I've noticed that, maybe there is no proper test for parameter order in the testsuite.  As unlike xopCmp (which is [where I deviate in the codebase](https://github.com/gcc-mirror/gcc/blob/e97cfaa9f61429f6717fd35d73309b02f9c02e95/gcc/d/dmd/clone.d#L701-L704)) getting the order wrong _usually_ won't affect the outcome for equality.

To unpick what's going on here, I'll ignore `p` and `q` and just name all variables `other` and `self` (self being the `this` reference).

For opCmp/compare:

1. User code calls `ti.compare(void* self, void* other)`
2. TypeInfo.compare calls `(*xopCmp)(other, self)`, which can result in two outcomes:
  - `xopCmp` is a pointer to member function `opCmp(const ref T other)` (for DMD the `this` parameter is `self`).
  - `xopCmp` is this generated function `xopCmp(T other, T self)`, and it calls `self.opCmp(other)`

For opEquals/equals, you're right though, for whatever reason, it's the reverse.

1. User code calls `ti.equals(void* self, void* other)`
2. TypeInfo.equals calls `(*xopEquals)(self, other)`, which can result in two outcomes:
  - `xopEquals` is a pointer to member function `opEquals(const ref T self)`, (for DMD the `this` parameter is `other`).
  - `xopEquals` is this generated function `xopEquals(T self, T other)`, and it returns `self.opEquals(other)`).

Looks like we need to swap the parameters in druntime, and return `this == p` here...

Reply:
> Have you tried to simplify this without any swapping and then hit issues? I'm asking because I suspect this weirdness to be the culprit for remaining issues in ldc-developers/ldc#3873 - the whole testsuite is green, but a compiler built itself with the new ABI then miscompiles, complaining about template forward references IIRC. I suspect something in the DMD frontend src to depend on this swapped opEquals order.

The template forward references are due to druntime calling `xopEquals` with the `other` parameter in the `this` position as I've outlined above, what you need to do is [swap them in druntime](https://github.com/gcc-mirror/gcc/blob/263a5944fc806396ecc3eff3d96277602e88ae2b/libphobos/libdruntime/object.d#L1742-L1747) too.

Unfortunately, [existing code already expects it to be this way](https://github.com/dlang/dmd/blob/5436d4d167e41f59b799071d8136bb051c87ae56/src/dmd/dtemplate.d#L7806-L7811) (otherwise the call would be `ti.equalsx(s.ti)`).

Reply:
Confirmed, and bug has been filed... _Sigh_ this is _really bad_ - thanks for making it more obvious @kinke.

https://issues.dlang.org/show_bug.cgi?id=22717

Reply:
> what you need to do is swap them in druntime too.

@kinke however, [this PR](https://github.com/dlang/druntime/commit/348a4ff50eabe62e0db5859c88cb0cbef08aace3) that recently got merged in means that you'll no longer have to hack around druntime to ensure the order is passed correctly.

Reply:
Excellent findings, thx for digging! - Yeah, pretty bad...

Reply:
> Excellent findings, thx for digging! - Yeah, pretty bad...

To unpick the call stack a bit more...
[`key in aa`](https://github.com/dlang/dmd/blob/5436d4d167e41f59b799071d8136bb051c87ae56/src/dmd/dtemplate.d#L2473)
-> [`keyti.equals(key, aaentry)`](https://github.com/dlang/druntime/blob/348a4ff50eabe62e0db5859c88cb0cbef08aace3/src/rt/aaA.d#L120)
  -> [`aaentry.opEquals(key)`](https://github.com/dlang/druntime/blob/348a4ff50eabe62e0db5859c88cb0cbef08aace3/src/object.d#L1888-L1889)
    -> [`key.equalsx(aaentry)`](https://github.com/dlang/dmd/blob/5436d4d167e41f59b799071d8136bb051c87ae56/src/dmd/dtemplate.d#L7810)
      -> [`aaentry.toAlias()`](https://github.com/dlang/dmd/blob/5436d4d167e41f59b799071d8136bb051c87ae56/src/dmd/dtemplate.d#L6110)

Swap it around, and it's the `key` that has `toAlias()` called on it, which will result in the forward referenced errors you'll see, because the "inflight" template instantiation is incomplete.